drivers/dp/swdp_bitbang.c

```diff
@@ -0,0 +1,632 @@
+
+static int sw_transfer(const struct device *dev,
+		       const uint8_t request, uint32_t *const data,
+		       const uint8_t idle_cycles, uint8_t *const response)
+{
+	struct sw_cfg_data *sw_data = dev->data;
+	unsigned int key;
+	uint32_t ack;
+	uint32_t bit;
+	uint32_t val;
+	uint32_t parity = 0;
+	uint32_t n;
+
+	LOG_DBG("request 0x%02x idle %u", request, idle_cycles);
+	if (!(request & SWDP_REQUEST_RnW)) {
+		LOG_DBG("write data 0x%08x", *data);
+		parity = sw_get32bit_parity(*data);
+	}
+
+	key = irq_lock();
+
+	val = sw_request_lut[request & 0xFU];
+	for (n = 8U; n; n--) {
+		SW_WRITE_BIT(dev, val, sw_data->clock_delay);
+		val >>= 1;
+	}
+
+	pin_swdio_out_disable(dev);
+	sw_cycle_turnaround(dev);
+
+	/* Acknowledge response */
+	SW_READ_BIT(dev, bit, sw_data->clock_delay);
+	ack = bit << 0;
+	SW_READ_BIT(dev, bit, sw_data->clock_delay);
+	ack |= bit << 1;
+	SW_READ_BIT(dev, bit, sw_data->clock_delay);
+	ack |= bit << 2;
+
+	if (ack == SWDP_ACK_OK) {
+		/* Data transfer */
+		if (request & SWDP_REQUEST_RnW) {
+			/* Read data */
+			val = 0U;
+			for (n = 32U; n; n--) {
+				/* Read RDATA[0:31] */
+				SW_READ_BIT(dev, bit, sw_data->clock_delay);
+				val >>= 1;
+				val |= bit << 31;
+			}
+
+			/* Read parity bit */
+			SW_READ_BIT(dev, bit, sw_data->clock_delay);
+			sw_cycle_turnaround(dev);
+			pin_swdio_out_enable(dev);
+
+			if ((sw_get32bit_parity(val) ^ bit) & 1U) {
+				ack = SWDP_TRANSFER_ERROR;
+			}
+
+			if (data) {
+				*data = val;
+			}
+
+		} else {
+			sw_cycle_turnaround(dev);
+
+			pin_swdio_out_enable(dev);
+			/* Write data */
+			val = *data;
+			for (n = 32U; n; n--) {
+				SW_WRITE_BIT(dev, val, sw_data->clock_delay);
+				val >>= 1;
+			}
+
+			/* Write parity bit */
+			SW_WRITE_BIT(dev, parity, sw_data->clock_delay);
+		}
+		/* Idle cycles */
+		n = idle_cycles;
+		if (n) {
+			pin_swdio_out(dev, 0U);
+			for (; n; n--) {
+				SW_CLOCK_CYCLE(dev, sw_data->clock_delay);
+			}
+		}
+
+		pin_swdio_out(dev, 1U);
+		irq_unlock(key);
+		if (request & SWDP_REQUEST_RnW) {
+			LOG_DBG("read data 0x%08x", *data);
+		}
+
+		if (response) {
+			*response = (uint8_t)ack;
+		}
+
+		return 0;
+	}
+
+	if ((ack == SWDP_ACK_WAIT) || (ack == SWDP_ACK_FAULT)) {
+		/* WAIT OR fault response */
+		if (sw_data->data_phase) {
+			for (n = 32U + 1U + sw_data->turnaround; n; n--) {
+				/* Dummy Read RDATA[0:31] + Parity */
+				SW_CLOCK_CYCLE(dev, sw_data->clock_delay);
+			}
+		} else {
+			sw_cycle_turnaround(dev);
+		}
+
+		pin_swdio_out_enable(dev);
+		pin_swdio_out(dev, 1U);
+		irq_unlock(key);
+		LOG_DBG("Transfer wait or fault");
+		if (response) {
+			*response = (uint8_t)ack;
+		}
+
+		return 0;
+	}
+
+	/* Protocol error */
+	for (n = sw_data->turnaround + 32U + 1U; n; n--) {
+		/* Back off data phase */
+		SW_CLOCK_CYCLE(dev, sw_data->clock_delay);
+	}
+
+	pin_swdio_out_enable(dev);
+	pin_swdio_out(dev, 1U);
+	irq_unlock(key);
+	LOG_INF("Protocol error");
+	if (response) {
+		*response = (uint8_t)ack;
+	}
+
+	return 0;
```

Comment:
Should  `0` be the result in case of a protocol error?

Reply:
Yes, protocol error is handled using ack/response, so we can use return value for controller IO errors.